File: backend-springboot/src/main/java/com/eventplanner/dto/EventResponse.java
package com.eventplanner.dto;

import java.time.LocalDateTime;
import java.util.Set;

public class EventResponse {

    private Long id;
    private String title;
    private String status;
    private LocalDateTime startTime;
    private LocalDateTime endTime;
    private Integer capacity;
    private Integer registeredCount;
    private Long organizerId;
    private String organizerName;
    private Long categoryId;
    private String categoryName;
    private Long venueId;
    private String venueName;
    private Set<String> tags;
    private LocalDateTime createdAt;

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public String getTitle() { return title; }
    public void setTitle(String title) { this.title = title; }
    public String getStatus() { return status; }
    public void setStatus(String status) { this.status = status; }
    public LocalDateTime getStartTime() { return startTime; }
    public void setStartTime(LocalDateTime startTime) { this.startTime = startTime; }
    public LocalDateTime getEndTime() { return endTime; }
    public void setEndTime(LocalDateTime endTime) { this.endTime = endTime; }
    public Integer getCapacity() { return capacity; }
    public void setCapacity(Integer capacity) { this.capacity = capacity; }
    public Integer getRegisteredCount() { return registeredCount; }
    public void setRegisteredCount(Integer registeredCount) { this.registeredCount = registeredCount; }
    public Long getOrganizerId() { return organizerId; }
    public void setOrganizerId(Long organizerId) { this.organizerId = organizerId; }
    public String getOrganizerName() { return organizerName; }
    public void setOrganizerName(String organizerName) { this.organizerName = organizerName; }
    public Long getCategoryId() { return categoryId; }
    public void setCategoryId(Long categoryId) { this.categoryId = categoryId; }
    public String getCategoryName() { return categoryName; }
    public void setCategoryName(String categoryName) { this.categoryName = categoryName; }
    public Long getVenueId() { return venueId; }
    public void setVenueId(Long venueId) { this.venueId = venueId; }
    public String getVenueName() { return venueName; }
    public void setVenueName(String venueName) { this.venueName = venueName; }
    public Set<String> getTags() { return tags; }
    public void setTags(Set<String> tags) { this.tags = tags; }
    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
}

File: backend-nodejs/src/helpers/eventResponse.js
const User = require('../models/User');
const Category = require('../models/Category');
const Venue = require('../models/Venue');
const Tag = require('../models/Tag');
const Registration = require('../models/Registration');

async function toEventResponse(event) {
  const [organizer, category, venue, tags, registeredCount] = await Promise.all([
    User.findOne({ id: event.organizerId }),
    event.categoryId != null ? Category.findOne({ id: event.categoryId }) : null,
    event.venueId != null ? Venue.findOne({ id: event.venueId }) : null,
    event.tagIds && event.tagIds.length ? Tag.find({ id: { $in: event.tagIds } }) : [],
    Registration.countDocuments({ eventId: event.id, status: 'CONFIRMED' })
  ]);

  return {
    id: event.id,
    title: event.title,
    status: event.status,
    startTime: event.startTime,
    endTime: event.endTime,
    capacity: event.capacity,
    registeredCount,
    organizerId: event.organizerId,
    organizerName: organizer ? organizer.name : null,
    categoryId: event.categoryId || null,
    categoryName: category ? category.name : null,
    venueId: event.venueId || null,
    venueName: venue ? venue.name : null,
    tags: tags.map((t) => t.name),
    createdAt: event.createdAt
  };
}

async function toEventResponseList(events) {
  return Promise.all(events.map(toEventResponse));
}

module.exports = { toEventResponse, toEventResponseList };
